File: LeetcodeJava/src/main/java/org/khwu/LE206ReverseList.java
package org.khwu;

import utility.ListNode;

public class LE206ReverseList {
    // MY
//    public ListNode reverseList(ListNode head) {
//        if (head == null) {
//            return head;
//        }
//        ListNode dummy = new ListNode(0);
//        dummy.next = head;
//        ListNode end = head;
//        ListNode swap = head.next;
//        while (swap != null) {
//            end.next = swap.next;
//            swap.next = dummy.next;
//            dummy.next = swap;
//            swap = end.next;
//        }
//        return dummy.next;
//    }

    // ref: https://discuss.leetcode.com/topic/16506/my-java-recursive-solution
    public ListNode reverseList(ListNode head) {
        if(head==null || head.next==null)
            return head;
        ListNode nextNode=head.next;
        ListNode newHead=reverseList(nextNode);
        nextNode.next=head;
        head.next=null;
        return newHead;
    }
}

File: LeetcodePython/LE206ReverseList.py
class SolutionIteration:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = ListNode(0)
        dummy.next = head
        node = head

        while node and node.next:
            tmp = node.next
            node.next = tmp.next
            tmp.next = dummy.next
            dummy.next = tmp

        return dummy.next


class SolutionRecursion:
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head

        next_node = head.next
        new_head = self.reverseList(head.next)
        next_node.next = head
        head.next = None
        return new_head